tls/extensions/s2n_server_renegotiation_info.c

```diff
@@ -69,12 +69,53 @@ static bool s2n_renegotiation_info_should_send(struct s2n_connection *conn)
  *#    include an empty "renegotiation_info" extension in the ServerHello
  *#    message.
  */
-static int s2n_renegotiation_info_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+static int s2n_renegotiation_info_send_initial(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
     POSIX_GUARD(s2n_stuffer_write_uint8(out, 0));
     return S2N_SUCCESS;
 }
 
+static int s2n_renegotiation_info_send_renegotiation(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    POSIX_ENSURE_REF(conn);
+
+    /* s2n-tls servers do not support renegotiation.
+     * We add the renegotiation version of this logic only for testing.
+     */
+    POSIX_ENSURE(s2n_in_unit_test(), S2N_ERR_NOT_IN_UNIT_TEST);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5746#3.7
+     *# This text applies if the connection's "secure_renegotiation" flag is
+     *# set to TRUE (if it is set to FALSE, see Section 4.4).
+     */
+    POSIX_ENSURE(conn->secure_renegotiation, S2N_ERR_NO_RENEGOTIATION);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5746#3.7
+     *# o  The server MUST include a "renegotiation_info" extension
+     *#    containing the saved client_verify_data and server_verify_data in
+     *#    the ServerHello.
+     */
+    const uint8_t verify_data_len = conn->handshake.finished_len;
+    POSIX_ENSURE_GT(verify_data_len, 0);
+    POSIX_GUARD(s2n_stuffer_write_uint8(out, verify_data_len * 2));
+    POSIX_GUARD(s2n_stuffer_write_bytes(out, conn->handshake.client_finished, verify_data_len));
+    POSIX_GUARD(s2n_stuffer_write_bytes(out, conn->handshake.server_finished, verify_data_len));
```

Comment:
The RFC says this will write a set length depending on the version:
```
For ServerHellos that are renegotiating, this field contains the
concatenation of client_verify_data and server_verify_data.  For
current versions of TLS, this will be a 24-byte value (for SSLv3,
it will be a 72-byte value).
```

Will `verify_data_len` always be the correct length? From a comment it seems like it changes: https://github.com/aws/s2n-tls/blob/37064dce11dedc3f505557531e92315eeda91173/tls/s2n_handshake.h#L152

Reply:
The answer to that is kind of complicated. It varies by cipher suite in TLS1.3 (but that's not relevant here, since this should never trigger for TLS1.3), is one constant value in TLS1.2, and is another constant value in SSLv3. That's the reason I did this work earlier: https://github.com/aws/s2n-tls/pull/3494

If it's not the correct length, the connection will fail, which is the correct response.

Reply:
Got it okay. Thanks!